android/expoview/src/main/java/host/exp/exponent/notifications/schedulers/CalendarScheduler.java

```diff
@@ -0,0 +1,225 @@
+
+  @Column
+  boolean scheduled = false;
+
+  @Column
+  String serializedDetails;
+
+  @Column
+  String calendarData;
+
+  // -- scheduler methods --
+
+  @Override
+  public void setApplicationContext(Context context) {
+    mApplicationContext = context.getApplicationContext();
+  }
+
+  @Override
+  public void schedule(String action) throws UnableToScheduleException {
+    if (!mTriggeringActions.contains(action)) {
+      return;
+    }
+    long nextAppearanceTime = 0;
+
+    try {
+      nextAppearanceTime = getNextAppearanceTime();
+    } catch (IllegalArgumentException e) {
+      throw new UnableToScheduleException();
+    }
+
+    ensureDetails();
+    try {
+      getManager().schedule(experienceId, notificationId, details, nextAppearanceTime, null);
+    } catch (ClassNotFoundException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public String getIdAsString() {
+    return Integer.valueOf(id).toString();
+  }
+
+  @Override
+  public String getOwnerExperienceId() {
+    return experienceId;
+  }
+
+  private void ensureDetails() {
+    try {
+      details = HashMapSerializer.deserialize(serializedDetails);
+    } catch (JSONException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void cancel() {
+    getManager().cancel(experienceId, notificationId);
+  }
+
+  @Override
+  public boolean canBeRescheduled() {
+    return repeat || (!scheduled);
+  }
+
+  @Override
+  public void onPostSchedule() {
+    this.scheduled = true;
+    save();
+  }
+
+  @Override
+  public String saveAndGetId() {
+    details.put(SchedulersManagerProxy.SCHEDULER_ID, getIdAsString());
+    setDetails(details);
+    save();
+    return getIdAsString();
+  }
+
+  @Override
+  public void remove() {
+    cancel();
+    delete();
+  }
+
+  private long getNextAppearanceTime() { // elapsedTime
+    CronDefinition cronDefinition = ExpoCronDefinitionBuilder.getCronDefinition();
+    CronParser parser = new CronParser(cronDefinition);
+    Cron cron = parser.parse(calendarData);
+
+    DateTime now = DateTime.now();
+    DateTime nextExecution = ExecutionTime.forCron(cron).nextExecution(now);
+    long whenShouldAppear = nextExecution.toDate().getTime();
+    long bootTime = DateTime.now().toDate().getTime() - SystemClock.elapsedRealtime();
+    return whenShouldAppear - bootTime;
+  }
+
+  private ExponentNotificationManager getManager() {
+    return new ExponentNotificationManager(mApplicationContext);
+  }
+
+  // model getters and setters
+
+  public int getId() {
+    return id;
+  }
+
+  public void setId(int id) {
+    this.id = id;
+  }
+
+  public int getNotificationId() {
+    return notificationId;
+  }
+
+  public void setNotificationId(int notificationId) {
+    this.notificationId = notificationId;
+  }
+
+  public String getExperienceId() {
+    return experienceId;
+  }
+
+  public void setExperienceId(String experienceId) {
+    this.experienceId = experienceId;
+  }
+
+  public boolean isRepeat() {
```

Comment:
A lot of these methods are shared between `CalendarScheduler` and `Timer?Scheduler`. Is it possible to somehow coalesce them into one base `BaseModel` and only override the `getNextAppearanceTime`?